### PR TITLE
Redis user session

### PR DESCRIPTION
**Pull request by happymancoder**
Merged after #5 

* Added implementation of persistent redis store for `IUserSessionStore`

### CODE REVIEW

File: src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
+﻿namespace ClickView.GoodStuff.AspNetCore.Authentication.StackExchangeRedis
```
Comment (MrSmoke) [inline src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs:1]: would it be possible to move this into its own project? just so if we end up using something like MySql for session storage in a project we dont need to install all the redis packages?

File: src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    using System.Threading.Tasks;
+
+    public class RedisUserSessionStore : IUserSessionStore
+    {
+        private readonly IDatabase _database;
+        private readonly TimeSpan? _expiry;
+
+        public RedisUserSessionStore(IOptions<RedisUserSessionCacheOptions> cacheOptions)
+        {
+            var options = cacheOptions.Value;
```
Comment (MrSmoke) [inline src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs:18]: if your testing options for null below, might also be worth checking `cacheOptions` for null

File: src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+        public async Task<UserSession?> GetInternalAsync(string key, CancellationToken token = default)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            token.ThrowIfCancellationRequested();
+
+            var session = await _database.StringGetAsync(key);
+            return session.HasValue ? JsonConvert.DeserializeObject<UserSession>(session.ToString()) : null;
```
Comment (MrSmoke) [inline src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs:92]: could we use `System.Text.Json` instead? I don't think we need full blown newtonsoft, and it also be serialize to a byte array which works nicely with redis

File: src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        private Task AddSessionIdAsync(UserSession? session, CancellationToken token = default)
+        {
+            if (string.IsNullOrWhiteSpace(session?.SessionId))
+                return Task.CompletedTask;
+
+            token.ThrowIfCancellationRequested();
+
+            return _database.StringSetAsync(GetSessionIdKey(session.SessionId), session.Key, _expiry);
```
Comment (MrSmoke) [inline src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs:113]: ah in #7 i added expire times to the session (forgot) so this should probably use that instead

**Review by MrSmoke — CHANGES_REQUESTED** [review]

File: src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs (hunk 4 above)
Comment (happymancoder) [reply src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs]: done.

File: src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs (hunk 3 above)
Comment (happymancoder) [reply src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs]: done.

File: src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs (hunk 2 above)
Comment (happymancoder) [reply src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs]: done.

File: src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs (hunk 1 above)
Comment (happymancoder) [reply src/AspNetCore/Authentication/Authentication/src/StackExchangeRedis/RedisUserSessionStore.cs]: done.

**Review by MrSmoke — APPROVED** [review]

**Pull request merged**